File: mac_gui/send_to_windows/Sources/TransferCore/Checksum.swift
// Checksum.swift — §4.6 sidecar 포맷.
//
// shasum -a 256 / sha256sum -c 호환:
//   <hex(lowercase)>  <filename>\n
// 헥스 두 칸 공백 파일명, LF 줄바꿈.
//
// 파일 모드: 1 line.
// 폴더 모드: 폴더 내부 모든 파일 라인 + 마지막에 combined dir-hash + "# combined dir-hash" 코멘트.
// 파일명은 NFC 정규화 (HFS+/SMB NFD 회피).

import Foundation

public enum Checksum {

    /// 파일 모드 sidecar 텍스트.
    public static func renderFile(sha256: String, filename: String) -> String {
        let nfc = filename.precomposedStringWithCanonicalMapping
        return "\(sha256)  \(nfc)\n"
    }

    /// 폴더 모드 sidecar 텍스트.
    /// - parameter folderName: 도착 후 폴더의 최종 이름 (네이밍 적용된).
    /// - parameter digest: Hashing.dirHash 결과.
    public static func renderDirectory(folderName: String,
                                       digest: Hashing.DirectoryDigest) -> String {
        let nfcFolder = folderName.precomposedStringWithCanonicalMapping
        var out = ""
        for ent in digest.entries {
            let nfcRel = ent.relativePath.precomposedStringWithCanonicalMapping
            out += "\(ent.sha256)  \(nfcFolder)/\(nfcRel)\n"
        }
        out += "\(digest.combined)  \(nfcFolder)  # combined dir-hash\n"
        return out
    }
}

File: mac_gui/share-manager/src-tauri/src/transfer/checksum.rs
// checksum.rs — port of Checksum.swift. §4.6 sidecar format.
//
// shasum -a 256 / sha256sum -c compatible:
//   "<hex>  <name>\n"  (two-space separator, LF line endings)
//
// Folder mode lists every contained file plus a final combined dir-hash
// line annotated with "# combined dir-hash". The Windows-side comment is
// optional metadata; shareguard verify treats `#`-prefixed lines as
// comments.

use unicode_normalization::UnicodeNormalization;

use super::hashing::DirectoryDigest;

pub fn render_file(sha256: &str, filename: &str) -> String {
    let nfc: String = filename.nfc().collect();
    format!("{sha256}  {nfc}\n")
}

pub fn render_directory(folder_name: &str, digest: &DirectoryDigest) -> String {
    let nfc_folder: String = folder_name.nfc().collect();
    let mut out = String::new();
    for e in &digest.entries {
        let nfc_rel: String = e.relative_path.nfc().collect();
        out.push_str(&format!("{}  {nfc_folder}/{nfc_rel}\n", e.sha256));
    }
    out.push_str(&format!(
        "{}  {nfc_folder}  # combined dir-hash\n",
        digest.combined
    ));
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn file_line_uses_two_space_separator() {
        let line = render_file("abc", "foo.txt");
        assert_eq!(line, "abc  foo.txt\n");
    }
}
